File: telldus-gui/Plugins/Live/livemessage.cpp
#include "livemessage.h"
#include "livemessagetoken.h"

#include <QVariant>
#include <QDebug>

class LiveMessage::PrivateData {
public:
	QList<LiveMessageToken> args;
};

LiveMessage::LiveMessage(const QString &name )
		: QObject()
{
	d = new PrivateData;
	this->append(name);
}

LiveMessage::LiveMessage(const QList<LiveMessageToken> &args )
		: QObject()
{
	d = new PrivateData;
	d->args = args;
}

LiveMessage::~LiveMessage() {
	delete d;
}

void LiveMessage::append(const QString &argument) {
	this->append(LiveMessageToken(argument));
}

void LiveMessage::append(const LiveMessageToken &argument) {
	d->args.append(argument);
}

void LiveMessage::appendToken(LiveMessageTokenScriptWrapper *argument) {
	this->append(argument->token());
}

LiveMessageToken LiveMessage::arg(int index) const {
	if (index + 1 >= d->args.count()) {
		return "";
	}
	return d->args.at(index +1 );
}

LiveMessageTokenScriptWrapper *LiveMessage::argument(int index) const {
	return new LiveMessageTokenScriptWrapper(arg(index));
}


QString LiveMessage::name() const {
	return this->arg(-1).stringVal.toLower();
}

QByteArray LiveMessage::toByteArray() const {
	QByteArray retval;
	foreach(LiveMessageToken arg, d->args) {
		retval.append(arg.toByteArray());
	}
	return retval;
}


LiveMessage LiveMessage::fromByteArray(const QByteArray &ba) {
	int start = 0;
	QList<LiveMessageToken> list;
	while (start < ba.length()) {
		LiveMessageToken d = LiveMessageToken::parseToken(ba, &start);
		if (d.valueType == LiveMessageToken::Invalid) {
			break;
		}
		list.append(d);
	}
	return LiveMessage(list);
}

File: examples/python/live/server/LiveMessage.py
# -*- coding: utf-8 -*-

import hashlib
from LiveMessageToken import *

class LiveMessage():
	def __init__(self, name = ""):
		if (name != ""):
			self.args = [LiveMessageToken(name)]
		else:
			self.args = []

	def append(self, argument):
		self.args.append(LiveMessageToken(argument))

	def argument(self,index):
		if (len(self.args) > index+1):
			return self.args[index+1]

		return LiveMessageToken()

	def count(self):
		return len(self.args)-1

	def name(self):
		return self.argument(-1).stringVal.lower()

	def toByteArray(self):
		retval = ''
		for arg in self.args:
			retval = retval + arg.toByteArray()
		return retval

	def toSignedMessage(self, hashMethod, privateKey):
		message = self.toByteArray()
		envelope = LiveMessage(LiveMessage.signatureForMessage(message, hashMethod, privateKey))
		envelope.append(message)
		return envelope.toByteArray()

	def verifySignature(self, hashMethod, privateKey):
		signature = self.name()
		rawMessage = self.argument(0).stringVal
		return (self.signatureForMessage(rawMessage, hashMethod, privateKey) == signature)

	@staticmethod
	def fromByteArray(rawString):
		list = []
		start = 0
		while (start < len(rawString)):
			start, token = LiveMessageToken.parseToken(rawString, start)
			if (token.valueType == LiveMessageToken.TYPE_INVALID):
				break
			list.append(token)

		msg = LiveMessage()
		msg.args = list
		return msg

	@staticmethod
	def signatureForMessage(msg, hashMethod, privateKey):
		h = 0
		if (hashMethod == "sha512"):
			h = hashlib.sha512()
		elif (hashMethod == "sha256"):
			h = hashlib.sha256()
		else:
			h = hashlib.sha1()

		h.update(msg)
		h.update(privateKey)
		return h.hexdigest().lower()
